src/translate.py

```diff
@@ -459,60 +461,71 @@ def update_min(old_or_none: Optional[int], new: int) -> int:
             # Initial saving of callee-save register onto the stack.
             stack_offset = inst.args[1].lhs_as_literal()
             info.callee_save_reg_locations[destination] = stack_offset
-            if inst.mnemonic == "sdc1":
-                callee_saved_doubles.append(stack_offset)
-            else:
-                callee_saved_words.append(stack_offset)
-        elif (
-            inst.mnemonic in ["lw", "lwc1", "ldc1"]
-            and isinstance(inst.args[1], AsmAddressMode)
-            and inst.args[1].rhs.register_name == "sp"
-        ):
-            info.subroutine_arg_top = update_min(
-                info.subroutine_arg_top, inst.args[1].lhs_as_literal()
-            )
-        elif (
-            inst.mnemonic == "addiu"
-            and inst.args[1] == Register("sp")
-            and isinstance(inst.args[2], AsmLiteral)
-            and inst.args[2].value < info.allocated_stack_size
-        ):
-            info.subroutine_arg_top = update_min(
-                info.subroutine_arg_top, inst.args[2].value
+            callee_saved_offset_and_size.append(
+                (stack_offset, 8 if inst.mnemonic == "sdc1" else 4)
             )
 
+    # Iterate over the whole function, not just the first basic block,
+    # to estimate the boundary for the subroutine argument region
+    for node in flow_graph.nodes:
+        for inst in node.block.instructions:
+            if not inst.args or not isinstance(inst.args[0], Register):
+                continue
+            destination = inst.args[0]
+
+            if (
+                inst.mnemonic in ["lw", "lwc1", "ldc1"]
+                and isinstance(inst.args[1], AsmAddressMode)
+                and inst.args[1].rhs.register_name == "sp"
+                and inst.args[1].lhs_as_literal() > 16
+            ):
+                info.subroutine_arg_top = update_min(
+                    info.subroutine_arg_top, inst.args[1].lhs_as_literal()
+                )
```

Comment:
now that allocated_stack_size is computed in the loop above, we could initialize `subroutine_arg_top` to `allocated_stack_size` and make it non-Optional